server/src/main/java/org/opensearch/index/engine/NRTReplicationEngine.java

```diff
@@ -71,18 +72,50 @@ public NRTReplicationEngine(EngineConfig engineConfig) {
             this.completionStatsCache = new CompletionStatsCache(() -> acquireSearcher("completion_stats"));
             this.readerManager = readerManager;
             this.readerManager.addListener(completionStatsCache);
-            this.translog = openTranslog(
-                engineConfig,
-                getTranslogDeletionPolicy(engineConfig),
+            final Map<String, String> userData = store.readLastCommittedSegmentsInfo().getUserData();
+            final String translogUUID = Objects.requireNonNull(userData.get(Translog.TRANSLOG_UUID_KEY));
+            translogManagerRef = new WriteOnlyTranslogManager(
+                engineConfig.getTranslogConfig(),
+                engineConfig.getPrimaryTermSupplier(),
                 engineConfig.getGlobalCheckpointSupplier(),
-                localCheckpointTracker::markSeqNoAsPersisted
+                getTranslogDeletionPolicy(engineConfig),
+                shardId,
+                readLock,
+                this::getLocalCheckpointTracker,
+                translogUUID,
+                new TranslogEventListener() {
+                    @Override
+                    public void onFailure(String reason, Exception ex) {
+                        failEngine(reason, ex);
+                    }
+
+                    @Override
+                    public void onAfterTranslogSync() {
+                        try {
+                            translogManager.getTranslog(false).trimUnreferencedReaders();
+                        } catch (IOException ex) {
+                            throw new TranslogException(shardId, "failed to trim unreferenced translog readers", ex);
+                        }
+                    }
+                },
+                this
             );
+            this.translogManager = translogManagerRef;
         } catch (IOException e) {
-            IOUtils.closeWhileHandlingException(store::decRef, readerManager);
+            Translog translog = null;
+            if (translogManagerRef != null) {
```

Comment:
I am wondering if `TranslogManager` should implement `Closeable` as well since it manages closeable resource (like `Translog`), wdyt?

Reply:
Thats a good point however there are multiple places where the Engine is consuming the underlying `Translog` directly, which I plan on getting rid of in subsequent PRs as a part of moving Translog to a module(this PR would become too big to review otherwise). 

https://github.com/opensearch-project/OpenSearch/blob/304d830275fdbe5f8095a7285ca66266987c727a/server/src/main/java/org/opensearch/index/translog/TranslogManager.java#L102

Once complete I will make the suggested change

Reply:
Opened https://github.com/opensearch-project/OpenSearch/issues/3709